# Awesome Array Driver

----

In [1]:
import aad

help(aad)

Help on package aad:

NAME
    aad

PACKAGE CONTENTS
    B1530Lib
    extlibs (package)
    mcd

CLASSES
    builtins.object
        AwesomeArrayDriver
    
    class AwesomeArrayDriver(builtins.object)
     |  AwesomeArrayDriver(uc_pid=22336, visa_addr='GPIB0::18::INSTR')
     |  
     |  Awesome Array Driver
     |  
     |  ...
     |  Attributes
     |  ----------
     |  _mcd: mcd.MCDriver
     |          The low-level driver used for the µc
     |  
     |  _b1530: B1530Lib.B1530
     |          The driver used to control the B1530
     |  
     |  Methods defined here:
     |  
     |  __init__(self, uc_pid=22336, visa_addr='GPIB0::18::INSTR')
     |      Creates the driver.
     |      
     |      Details:
     |              It will search for the µc using the PID value 'DEFAULT_PID' or the one provided in argument.
     |              Takes the first found if many have the same PID.
     |              RAISE Exception if not found.
     |      
     |      Arguments:
     | 

---

In [2]:
print("---- List of available ports ----")
aad.print_ports()

print("---- List of available lab equipment ----")
aad.print_visa_dev()

---- List of available ports ----
COM4 - USB Serial Device (COM4) | PID:  22336
COM3 - Intel(R) Active Management Technology - SOL (COM3) | PID:  None
---- List of available lab equipment ----
ASRL3::INSTR
ASRL4::INSTR
GPIB0::INTFC
GPIB0::18::INSTR


The communication with the µc is done via `USB Serial Device` at `COM4`.  
The driver will search by default for a device with `PID` 22336.   

The B1500 is at the adress `GPIB0::18::INSTR`.

In [3]:
driver = aad.AwesomeArrayDriver()

print("The driver selected the port:", driver._mcd.ser.port)

The driver selected the port: COM4


If no device with such `PID` was found, an exception would be raised.  
You can then check the connection, or use `aad.print_ports` to find the correct `PID` and give it as an argument of the constructor:  
```
driver = aad.AwesomeArrayDriver(uc_pid=...)
```

If the B1500 address is wrong, an exception would be raised.  
You can then check the connection, or use `aad.print_visa_dev` to find the correct address and give it as an argument of the constructor:
```
driver = aad.AwesomeArrayDriver(visa_addr=...)
```

---  
(For now, the driver only configures the shift registers)  
A memristor location is given by three parameters:
* `col`: integer : Address of the column in the array
* `row`: integer : Address of the row in the array
* `bar`: boolean : Whether to select the memristor Rb (if True) or R (otherwise)  

Then, you can set the individual state of a memristor using the functions `set` or `reset`.  
You can also form the memristor first using the function `form`.  

For a more complete description, see the `help` output above.

----

In [4]:
# Sanity check of the shift registers for one particular configuration
sanity = driver.test_sr_sanity(col=0, row=0, bar=False, set=True)
for sr_id in aad.SR_LIST:
	for bit_id in range(64):
		# Usually, the condition would be "not ..." but for this test, the output is random and most registers will appear as not sane
		if sanity[sr_id][bit_id]:
			print(f"{str(sr_id)}'s bit #{bit_id} is sane")

In [5]:
driver.set(col=0, row=0, bar=False)
driver.reset(col=0, row=0, bar=True)